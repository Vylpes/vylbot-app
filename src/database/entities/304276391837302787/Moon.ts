import { Column, Entity, IsNull } from "typeorm";
import BaseEntity from "../../../contracts/BaseEntity";
import AppDataSource from "../../dataSources/appDataSource";

@Entity()
export default class Moon extends BaseEntity {
    constructor(moonNumber: number, description: string, userId: string) {
        super();

        this.MoonNumber = moonNumber;
        this.Description = description;
        this.UserId = userId;
    }

    @Column()
    MoonNumber: number;

    @Column()
    Description: string;

    @Column({ nullable: true })
    WhenArchived?: Date;

    @Column()
    UserId: string;

    public static async FetchMoonsByUserId(userId: string): Promise<Moon[] | null> {
        const repository = AppDataSource.getRepository(Moon);

        const all = await repository.find({ where: { UserId: userId } });

        return all;
    }

    public static async FetchPaginatedMoonsByUserId(userId: string, pageLength: number, page: number): Promise<[ Moon[], number ]> {
        const rangeStart = page * pageLength;

        const repository = AppDataSource.getRepository(Moon);

        const moons = await repository.findAndCount({
            where: { UserId: userId, WhenArchived: IsNull() },
            order: { MoonNumber: "ASC" },
            skip: rangeStart,
            take: pageLength,
        });

        return moons;
    }

    public static async FetchMoonCountByUserId(userId: string): Promise<number> {
        const repository = AppDataSource.getRepository(Moon);

        const count = await repository.count({ where: { UserId: userId } });

        return count;
    }
}
