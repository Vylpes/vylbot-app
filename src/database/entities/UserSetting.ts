import { Column, Entity} from "typeorm";
import AppDataSource from "../dataSources/appDataSource";
import BaseEntity from "../../contracts/BaseEntity";

@Entity()
export default class UserSetting extends BaseEntity {
    constructor(userId: string, key: string, value: string) {
        super();

        this.UserId = userId;
        this.Key = key;
        this.Value = value;
    }

    @Column()
    UserId: string;

    @Column()
    Key: string;

    @Column()
    Value: string;

    public UpdateValue(value: string) {
        this.Value = value;
    }

    public static async FetchOneByKey(userId: string, key: string, relations?: string[]): Promise<UserSetting | null> {
        const repository = AppDataSource.getRepository(UserSetting);

        const single = await repository.findOne({ where: { UserId: userId, Key: key }, relations: relations || {} });

        return single;
    }
}
