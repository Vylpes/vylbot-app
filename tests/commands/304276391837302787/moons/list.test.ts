import { ActionRowBuilder, APIEmbed, ButtonBuilder, CommandInteraction, EmbedBuilder, InteractionReplyOptions } from "discord.js";
import List from "../../../../src/commands/304276391837302787/moons/list";
import Moon from "../../../../src/database/entities/304276391837302787/Moon";
import UserSetting from "../../../../src/database/entities/UserSetting";

describe("GIVEN happy flow", () => {
    let repliedWith: InteractionReplyOptions | undefined;

    const interaction = {
        reply: jest.fn((options) => {
            repliedWith = options;
        }),
        options: {
            get: jest.fn()
                .mockReturnValueOnce(undefined) // User
                .mockReturnValue({
                    value: "0",
                }), // Page
        },
        user: {
            id: "userId",
        }
    } as unknown as CommandInteraction;

    beforeAll(async () => {
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [
                {
                    MoonNumber: 1,
                    Description: "Test Description",
                }
            ],
            1,
        ]);

        UserSetting.FetchOneByKey = jest.fn().mockResolvedValue({
            Value: "0",
        });

        await List(interaction);
    });

    test("EXPECT interaction to be replied", () => {
        expect(interaction.reply).toHaveBeenCalledTimes(1);
        expect(repliedWith).toMatchSnapshot();
    });
});

describe("GIVEN moons returned is empty", () => {
    let repliedWith: InteractionReplyOptions | undefined;

    const interaction = {
        reply: jest.fn((options) => {
            repliedWith = options;
        }),
        options: {
            get: jest.fn()
                .mockReturnValueOnce(undefined) // User
                .mockReturnValue({
                    value: "0",
                }), // Page
        },
        user: {
            id: "userId",
        }
    } as unknown as CommandInteraction;

    beforeAll(async () => {
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [],
            0,
        ]);

        UserSetting.FetchOneByKey = jest.fn().mockResolvedValue({
            Value: "0",
        });

        await List(interaction);
    });

    test("EXPECT none description", () => {
        expect(repliedWith).toBeDefined();

        expect(repliedWith!.embeds).toBeDefined();
        expect(repliedWith!.embeds!.length).toBe(1);

        const repliedWithEmbed = repliedWith!.embeds![0] as EmbedBuilder;

        expect(repliedWithEmbed.data.description).toBe("*none*");
    });
});

describe("GIVEN it is the first page", () => {
    let repliedWith: InteractionReplyOptions | undefined;

    const interaction = {
        reply: jest.fn((options) => {
            repliedWith = options;
        }),
        options: {
            get: jest.fn()
                .mockReturnValueOnce(undefined) // User
                .mockReturnValue({
                    value: "0",
                }), // Page
        },
        user: {
            id: "userId",
        }
    } as unknown as CommandInteraction;

    beforeAll(async () => {
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [
                {
                    MoonNumber: 1,
                    Description: "Test Description",
                }
            ],
            1,
        ]);

        UserSetting.FetchOneByKey = jest.fn().mockResolvedValue({
            Value: "0",
        });

        await List(interaction);
    });

    test("EXPECT Previous button to be disabled", () => {
        expect(repliedWith).toBeDefined();

        expect(repliedWith!.components).toBeDefined();
        expect(repliedWith!.components!.length).toBe(1);

        const repliedWithRow = repliedWith!.components![0] as ActionRowBuilder<ButtonBuilder>;

        expect(repliedWithRow.components[0].data.disabled).toBe(true);
    });
});

describe("GIVEN it is the last page", () => {
    let repliedWith: InteractionReplyOptions | undefined;

    const interaction = {
        reply: jest.fn((options) => {
            repliedWith = options;
        }),
        options: {
            get: jest.fn()
                .mockReturnValueOnce(undefined) // User
                .mockReturnValue({
                    value: "0",
                }), // Page
        },
        user: {
            id: "userId",
        }
    } as unknown as CommandInteraction;

    beforeAll(async () => {
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [
                {
                    MoonNumber: 1,
                    Description: "Test Description",
                }
            ],
            1,
        ]);

        UserSetting.FetchOneByKey = jest.fn().mockResolvedValue({
            Value: "0",
        });

        await List(interaction);
    });

    test("EXPECT Next button to be disabled", () => {
        expect(repliedWith).toBeDefined();

        expect(repliedWith!.components).toBeDefined();
        expect(repliedWith!.components!.length).toBe(1);

        const repliedWithRow = repliedWith!.components![0] as ActionRowBuilder<ButtonBuilder>;

        expect(repliedWithRow.components[1].data.disabled).toBe(true);
    });

    describe("GIVEN moon count is greater than the amount of moons in the database", () => {
        beforeAll(async () => {
            UserSetting.FetchOneByKey = jest.fn().mockResolvedValue({
                Value: "2",
            });

            await List(interaction);
        });

        test("EXPECT untracked counter to be shown", () => {
            const repliedWithEmbed = repliedWith!.embeds![0] as EmbedBuilder;

            expect(repliedWithEmbed.data.description).toContain("...plus 1 more untracked");
        });
    });
});

describe("GIVEN moon count is empty", () => {
    let repliedWith: InteractionReplyOptions | undefined;

    const interaction = {
        reply: jest.fn((options) => {
            repliedWith = options;
        }),
        options: {
            get: jest.fn()
                .mockReturnValueOnce(undefined) // User
                .mockReturnValue({
                    value: "0",
                }), // Page
        },
        user: {
            id: "userId",
        }
    } as unknown as CommandInteraction;

    beforeAll(async () => {
        Moon.FetchPaginatedMoonsByUserId = jest.fn().mockResolvedValue([
            [],
            0,
        ]);

        UserSetting.FetchOneByKey = jest.fn().mockResolvedValue({
            Value: "0",
        });

        await List(interaction);
    });

    test("EXPECT Next button to be disabled", () => {
        expect(repliedWith).toBeDefined();

        expect(repliedWith!.components).toBeDefined();
        expect(repliedWith!.components!.length).toBe(1);

        const repliedWithRow = repliedWith!.components![0] as ActionRowBuilder<ButtonBuilder>;

        expect(repliedWithRow.components[1].data.disabled).toBe(true);
    });
});
