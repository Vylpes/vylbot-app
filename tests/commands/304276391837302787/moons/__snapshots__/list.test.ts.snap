// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GIVEN happy flow EXPECT interaction to be replied 1`] = `
{
  "components": [
    {
      "components": [
        {
          "custom_id": "moons list userId -1",
          "disabled": true,
          "emoji": undefined,
          "label": "Previous",
          "style": 1,
          "type": 2,
        },
        {
          "custom_id": "moons list userId 01",
          "disabled": true,
          "emoji": undefined,
          "label": "Next",
          "style": 1,
          "type": 2,
        },
      ],
      "type": 1,
    },
  ],
  "embeds": [
    {
      "color": 3166394,
      "description": "**1 -** Test Descriptio",
      "footer": {
        "icon_url": undefined,
        "text": "Page 01 of 1 Â· 0 moons",
      },
      "title": "undefined's Moons",
    },
  ],
}
`;
